### PR TITLE
Allow clients to set shape

### PR DESCRIPTION
**Pull request by copycat-killer**
*The squasher strikes back*

This is #1499 , which I accidentally deleted in my fork.

### CODE REVIEW

**Comment by Elv13** [comment]
You can run the tests locally if you have Xephyr installed. There is a "run.sh" script. In this case, `local gears =require("gears")` is missing

**Comment by copycat-killer** [comment]
There's an assertion fail with `test-awful-widget-button.lua`. Is it related?

**Comment by Elv13** [comment]
I don't see such assertion in the log. Maybe it happens locally on your system (or maybe you moved the mouse during the tests). WhatI see is

```

    lib/awful/client/shape.lua:55:11: shadowing upvalue shape on line 17

    lib/awful/client/shape.lua:102:23: shadowing upvalue shape on line 17

```

You can install `luacheck` and add it as a commit hook to spot those before they get into Travis (to save time)

**Comment by copycat-killer** [comment]
Travis shows this error which I can't understand:

```shell
== Running /home/travis/build/awesomeWM/awesome/tests/test-awful-client.lua ==
2017-02-01 21:31:49 W: awesome: a_glib_poll:389: Last main loop iteration took 1.015218 seconds! Increasing limit for this warning to that value.
2017-02-01 21:31:52 W: awesome: a_glib_poll:389: Last main loop iteration took 2.403711 seconds! Increasing limit for this warning to that value.
2017-02-01 21:32:08 W: awesome: a_glib_poll:389: Last main loop iteration took 2.995378 seconds! Increasing limit for this warning to that value.
2017-02-01 21:32:51 W: awesome: a_glib_poll:389: Last main loop iteration took 3.511933 seconds! Increasing limit for this warning to that value.
2017-02-01 21:33:22 W: awesome: a_glib_poll:389: Last main loop iteration took 5.520004 seconds! Increasing limit for this warning to that value.
2017-02-01 21:34:11 W: awesome: a_glib_poll:389: Last main loop iteration took 6.302974 seconds! Increasing limit for this warning to that value.
2017-02-01 21:34:35 W: awesome: a_glib_poll:389: Last main loop iteration took 6.333373 seconds! Increasing limit for this warning to that value.
Awesome was killed due to timeout after 180 seconds
===> ERROR running /home/travis/build/awesomeWM/awesome/tests/test-awful-client.lua <===
```

**Comment by Elv13** [comment]
Ignore that one and its sibling in test-awful-tag. The Travis node didn't have enough memory and started swapping. We have a hack to "fix" it in another PR, but it isn't merged yet (and is very ugly).

**Comment by copycat-killer** [comment]
This PR introduces `property::size`. What about introducing `property::position` for x,y as well?

**Comment by psychon** [comment]
Sure.
How about adding those new signals in a separate PR to get things in quicker and to have a smaller PR?

**Comment by copycat-killer** [comment]
Do we have to wait for the Travis fix before deciding on this PR, or is there something else I can do?

**Comment by psychon** [comment]
Travis failed this time because:
```
lib/awful/client/shape.lua:59:23: shadowing upvalue shape on line 17
```
Besides that, I don't have enough time to look much into this right now. I can take a look at the weekend, I think.

**Comment by copycat-killer** [comment]
Solved.

**Comment by Elv13** [comment]
This needs to be rebase. This time it may be easier to just re-apply the patch without the commit. Also make sure to merge the 2 code paths. XEye still has to work if no shape is set.

**Comment by actionless** [comment]
but this version of shape routine (`shape.update.all`) will change the shape only of an outer part of a client's border

i would recommend to use the later revision of that code, which can be found in my local config:
https://github.com/actionless/awesome_config/blob/devel/config/signals.lua#L25-L78

**Comment by Elv13** [comment]
@actionless the copycat version looks alright to me
pick8

**Comment by actionless** [comment]
how did you configured it?

i did only `c.shape = gears.shape.rounded_rect`:

2017-02-03--1486134400_1386x1330_scrot

**Comment by Elv13** [comment]
Uhh, I did the same (I added it to awful.rules). Does it only happens with GTK3 apps?

**Comment by Elv13** [comment]
Merged with some fixes by psychon and me

**Comment by actionless** [comment]
no, it happened the same with any apps, i will later on test the latest commit which got merged, mb there it will be fine for me

**Comment by actionless** [comment]
i think the problem what i am using 4 titlebars instead of borders

however the only reason why i am using titlebars there is because i can't subscribe to mouse clicks on a client's borders

so i am not even sure to which of two directions i should move with that

**Comment by psychon** [comment]
@actionless Did #1521 help with that? I think it should/might...

**Comment by actionless** [comment]
no, it didn't

**Comment by actionless** [comment]
so, let me rephrase my previous question:
should i try to fix mouse-event on a real client's border or should i update the current client_shape implementation to bring the changes from my local ones to support inner border shaping for the titlebars?

**Comment by psychon** [comment]
> should i try to fix mouse events on a real client's border

Feel free. However, X11 won't let you do that. Feel free to try with `xev` (that's what I did :-) ).

**Comment by psychon** [comment]
Today while showering I thought: This cannot work.

We apply the shape at the outside of the border and its inside. Since you have `border_width = 0`, the two are "at the same position".

For shaping titlebars: Each pixels can have one of three states:
* Outside of the window (not contained in the bounding shape)
* In the border (inside the bounding shape, outside the clip shape)
* Inside the window (inside of both shapes)

Since you have no border, there are no pixels inside of the border (and it wouldn't make sense to add a border "in the middle of the window"; Plus mouse clicks wouldn't be registered here). Of course, the area were you want the shape to be applied should not be outside of the window, because then the window would just have a hole. That leaves "inside the window" as the only alternative and this is just what we are doing.

**Comment by actionless** [comment]
but with the apply_shape code from my config it's working asexpected:

2017-02-07--1486472668_370x292_scrot

**Comment by psychon** [comment]
You have `border_width = 0` and use titlebars instead, right? What is visible in the "inside" of the rounded corner there? Put differently: How do the two shapes look like here?

**Comment by actionless** [comment]
yup, border width is 0, but it seems to be unrelated because i still can't reproduce the problem with any other border width

i think the main difference is about coordinates of the shape -- in my example the square is smaller, ie:

upstream:
`_shape(cr, geom.width + 2*c.border_width, geom.height + 2*c.border_width)`

mine:
`shape(cr, geo.width, geo.height, shape_args)`

**Comment by Elv13** [comment]
> shape(cr, geo.width, geo.height, shape_args)

For the shape_args part, it can be added later without breaking the API. If you open an issue for it, I can take care of it.

**Comment by actionless** [comment]
sorry, shape_args wasn't the point, the point was a smaller geometry

**Comment by Elv13** [comment]
I know, it was a side note. It's why I said to create an issue if you wan it
